<h1>ACTIVITY2: UK Accident Data</h1>
<hr>
<h2>Analyst: Anne Fedelino</h2>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h2>Filling Null Values & Filling Missing Values</h2>

In [4]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [31]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mean)
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean)
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mean)
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mean)
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mean)
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mean)

In [32]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [33]:
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident['Accident Date'] = accident['Accident Date'].str.replace('/', '-')

In [35]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

<h2>Adjusting Data Types</h2>

In [96]:
accident['Index'] = accident['Index'].astype('category')
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors='coerce')
accident['Latitude'] = accident['Latitude'].astype('float')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('float')
accident['Number_of_Casualties'] = accident['Number_of_Casualties'].astype('int')
accident['Number_of_Vehicles'] = accident['Number_of_Vehicles'].astype('int')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [98]:
accident.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
Year                                int32
Month                               int32
Day                                 int32
DayOfWeek                           int32
dtype: object

In [38]:
accident['Accident Date'] = accident['Accident Date'].fillna(accident['Accident Date'].mean())
accident['Year'] = accident['Year'].fillna(accident['Year'].mean())
accident['Month'] = accident['Month'].fillna(accident['Month'].mean())
accident['Day'] = accident['Day'].fillna(accident['Day'].mean())
accident['DayOfWeek'] = accident['DayOfWeek'].fillna(accident['DayOfWeek'].mean())

In [39]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

<h2>Extracting Date Information using Pandas Date Time</h3>

In [105]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek #moday=0, sunday=6

In [106]:
accident[['Year', 'Month', 'Day', 'DayOfWeek']].head()

Year  Month  Day  DayOfWeek
0  2019      6    5          2
1  2019      7    2          1
2  2019      8   26          0
3  2019      8   16          4
4  2019      9    3          1

<h2>Data Analytics</h2>
<h3>UniVariate</h3>
<h4>Summary of Accidents in 2019</h4>
<h4>Light Conditions</h4>
<hr>

<h2>Univariate Analysis on Accident that happened in 2019</h2>

In [44]:
df_2019 = accident[accident['Year'] == 2019]

In [45]:
print(f"total number of accident in 2019 is {df_2019['Light_Conditions'].value_counts().sum()}")
df_2019['Light_Conditions'].value_counts()

total number of accident in 2019 is 182115


Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

<h1>Data Analysis</h1>
<hr>

<h1>Q1. What is the total number of casualties for each accident severity?</h1>

In [46]:
accident.groupby("Accident_Severity")["Number_of_Casualties"].sum()

Accident_Severity
Fatal       16483
Serious    129439
Slight     750646
Name: Number_of_Casualties, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>It shows that there were 16,483 fatal accidents, 12,9439 serious accidents, and 750,646 minor accidents.</h3>

<h3>INSIGHT 2:</h3>
<h3>This shows that slight accidents have the highest number of casualties.</h3>

<h1>Q2. What is the maximum number of casualties in each accident severity category?</h1>

In [8]:
accident.groupby("Accident_Severity")["Number_of_Casualties"].max()

Accident_Severity
Fatal      68
Serious    45
Slight     47
Name: Number_of_Casualties, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>This shows that there were 68 fatal, 45 serious condition, and 47 in slight condition.</h3>

<h3>INSIGHT 2:</h3>
<h3>Fatal accidents show the maximum casualties per accident.</h3>

<h1>Q3. What are the types of vehicles involved in accidents?</h1>

In [116]:
accident['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>The result above shows the vehicles involved in accidents.</h3>

<h3>INSIGHT 2:</h3>
<h3>The car is the vehicle that has the most record involved in accidents.</h3>

<h1>Q4. How many accidents happened in each Urban_or_Rural_Area?</</h1>

In [10]:
accident["Urban_or_Rural_Area"].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>It shows that there were 421,678 vehicle accidents in urban areas, there are 238,990 vehicle accident in rural areas, and there are 11 accidents in unlocated areas.</h3>

<h3>INSIGHT 2:</h3>
<h3>Districts with high population density record more vehicles involved in accidents.</h3>

<h3>INSIGHT 3:</h3>
<h3>Urban areas often record more accidents than Rural areas.</h3>

<h1>Q5. What is the total number of casualties in urban vs rural areas?</h1>

In [11]:
accident.groupby("Urban_or_Rural_Area")["Number_of_Casualties"].sum()

Urban_or_Rural_Area
Rural          353515
Unallocated        13
Urban          543040
Name: Number_of_Casualties, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>The total number of casualties in rural areas is 353,515, in unlocated areas is 13, and there are 543,040 casualties in urban areas.</h3>

<h3>INSIGHT 2:</h3>
<h3>This shows that urban areas report more accidents than rural area.</h3>

<h1>Q6. How many accidents happened in each Road_Type?</h1>

In [12]:
accident["Road_Type"].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>There are 496663 accidents on 'Single Carriageways', 99424 accidents on 'Dual Carriageways', 43992 accidents on 'Roundabouts', 13559 accidents on 'One-way streets', and 7041 accidents on 'Slip roads'.</h3>

<h3>INSIGHT 2:</h3>
<h3>This shows that 'Single Carriageways dominate accidents by road types.</h3>

<h1>Q7. Which weather condition is most common in accidents?</h1>

In [111]:
accident["Weather_Conditions"].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>INSIGHT:</h3>
<h3>The result above shows that 'Fine no high winds' is the most common condition in accidents.</h3>

<h1>Q8. How manny serious accident are there in urban and rural area?</h1>

In [14]:
accident[accident["Accident_Severity"] == "Serious"]["Urban_or_Rural_Area"].value_counts()

Urban_or_Rural_Area
Urban          50904
Rural          37312
Unallocated        1
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>Thsere are 50,904 seious condition in urban areas, in rural areas there are 37,312 serious accident, and only 1 serious accident happened in an unllocated area.</h3>

<h3>INSIGHT 2:</h3>
<h3>This shows that Urban areas has more serious accident than rural(and unllocated area).</h3>


<h1>Q9. How many accidents of each severity happened under each light condition?</h1>

In [15]:
accident.groupby("Light_Conditions")["Accident_Severity"].value_counts()

Light_Conditions             Accident_Severity
Darkness - lighting unknown  Slight                 5622
                             Serious                 794
                             Fatal                    68
Darkness - lights lit        Slight               108345
                             Serious               19130
                             Fatal                  1860
Darkness - lights unlit      Slight                 2138
                             Serious                 360
                             Fatal                    45
Darkness - no lighting       Slight                28651
                             Serious                7174
                             Fatal                  1612
Daylight                     Slight               419045
                             Serious               60759
                             Fatal                  5076
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>Daylight condition have the highest number of accidents across all severity.</h3>

<h1>Q10. What is the maximum number of casualties in urban_or_rural areas?</h1>

In [114]:
accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].max()

Urban_or_Rural_Area
Rural          68
Unallocated     3
Urban          32
Name: Number_of_Casualties, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>The maximum number of casualties in Rural is 68, in Unllocated area is 3, and in Urban is 32.</h3>

<h3>INSIGHT 2:</h3>
<h3>This shows that Rural has the highest maximum number of casualties in all of the area.</h3>

<h1>Q11. How many fatal accidents are there in Sheffield?</h1>

In [19]:
accident[(accident["Accident_Severity"] == "Fatal") & (accident["District Area"] == "Sheffield")]["District Area"].value_counts()

District Area
Sheffield    38
Name: count, dtype: int64

<h3>INSIGHT:</h3>
<h3>There are 38 number of fatal accident that happened in Sheffield.</h3>

<h1>Q12. Which vehicle type is most often involved in accidents?</h1>

In [20]:
accident["Vehicle_Type"].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>According to the result, car is the most often involved in accidents.</h3>

<h3>INSIGHT 2:</h3>
<h3>THere are 16 vehicles in the list that are involved in accident, but car has the most number of involved in accident.</h3>

<h1>Q13. What is the maximum and minimum number of vehicles involved by severity?</h1>

In [119]:
accident.groupby('Accident_Severity')['Number_of_Vehicles'].max()

Accident_Severity
Fatal      28
Serious    19
Slight     32
Name: Number_of_Vehicles, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>There are 28 vehicles involved in fatal, 19 vehicles involved in serious, snd there are 32 vehicles that are involved in slight condition.</h3>

<h3>INSIGHT 2:</h3>
<h3>There are 32(maximum number) vehicles that are involved in slight severity.</h3>

<h3>INSIGHT 3:</h3>
<h3>There are 19(minimum number) vehicles that are involved in serious severity.</h3>

<h1>Q14. Which light condition has the most fatal accidents?</h1>

In [22]:
accident[accident["Accident_Severity"] == "Fatal"]["Light_Conditions"].value_counts()

Light_Conditions
Daylight                       5076
Darkness - lights lit          1860
Darkness - no lighting         1612
Darkness - lighting unknown      68
Darkness - lights unlit          45
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>Daylight has the highest/ most fatal accidents among the Accident Severity.</h3>

<h1>Q15. What is the minimum and maximum number of casualties recorded in an accident?</h1>

In [125]:
accident['Number_of_Casualties'].max()

np.int64(68)

In [126]:
accident['Number_of_Casualties'].min()

np.int64(1)

<h3>INSIGHT 1:</h3>
<h3>The maximum number of casualties is 68.</h3>

<h3>INSIGHT 2:</h3>
<h3>The minimum number of casualties is 1.</h3>

<h1>Q16. How many accidents happened in urban and rural areas by road surface condition?</h1>

In [24]:
accident.groupby("Urban_or_Rural_Area")["Road_Surface_Conditions"].value_counts()

Urban_or_Rural_Area  Road_Surface_Conditions
Rural                Dry                        145145
                     Wet or damp                 79005
                     Frost or ice                10953
                     Snow                         3102
                     Flood over 3cm. deep          785
Unallocated          Dry                             5
                     Wet or damp                     5
                     Flood over 3cm. deep            1
Urban                Dry                        303397
                     Wet or damp                107698
                     Frost or ice                 7564
                     Snow                         2788
                     Flood over 3cm. deep          231
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>In both urban and rural roads, most accidents happen on dry surfaces.</h3>

<h3>INSIGHT 2: </h3>
<h3>However, wet and icy roads in rural areas appear riskier due to higher speeds and less road infrastructure.</h3>

<h1>Q17. Which weather condition records the highest number of accidents?</h1>

In [25]:
accident["Weather_Conditions"].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>INSIGHT 1: </h3>
<h3>According to the result fine no high winds(clear) has the most  highest number of accidents.</h3>

<h3>INSIGHT 2:</h3>
<h3>Rainig no high winds and fine no high winds are the two weather with hig number of accident, but no high wind is the highest with 535,013 accidents.</h3>

<h1>Q18. Which road type has the most accidents?</h1>

In [26]:
accident["Road_Type"].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h3>INSIGHT: </h3>
<h3>Single carriageway has the most accidents in record with 496663 number of accidents.</h3>

<h1>Q19. Which accident severity level is most common?</h1>

In [50]:
accident["Accident_Severity"].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h3>INSIGHT 1:</h3>
<h3>Slight accidents is the most common accident severity level.</h3>

<h3>INSIGHT 2:</h3>
<h3>Among those three severity level/accident severity, slight has the highest number with 563801, serious accident has 88217, and fatal has 8661.</h3>

<h1>Q20. What is the correlation between number of vehicles and number of casualties?</h1>

In [51]:
accident[["Number_of_Vehicles", "Number_of_Casualties"]].corr()

Number_of_Vehicles  Number_of_Casualties
Number_of_Vehicles              1.000000              0.228889
Number_of_Casualties            0.228889              1.000000

<h1>Q21. Which year had the most accidents?</h1>

In [52]:
accident["Accident Date"].dt.year.value_counts()

Accident Date
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h1>Q22. Which month has the most accidents?</h1>

In [53]:
accident["Accident Date"].dt.month.value_counts()

Accident Date
11    60424
10    59580
7     57445
6     56481
9     56455
5     56352
3     54086
8     53913
1     52872
12    51836
4     51744
2     49491
Name: count, dtype: int64

<h1>Q23.Which year had the most accidents in urban areas?</h1>

In [87]:
accident[accident['Urban_or_Rural_Area'] == 'Urban'] ['Accident Date'].dt.year.value_counts()

Accident Date
2019    114552
2020    108594
2021    104538
2022     93994
Name: count, dtype: int64

<h1>Q24. Which month has the most accidents in rural areas?</h1>

In [129]:
accident[accident['Urban_or_Rural_Area'] == 'Rural'] ['Accident Date'].dt.month.value_counts()

Accident Date
11    21352
7     21215
8     20998
10    20786
6     20200
5     20125
9     20035
1     19540
12    19361
3     18806
4     18432
2     18140
Name: count, dtype: int64

<h1>Q24. Which day of week sees the most fatal accidents?</h1>

In [91]:
accident[accident['Accident_Severity'] == 'Fatal'] ['Accident Date'].dt.dayofweek.value_counts()

Accident Date
6    1522
0    1385
5    1326
2    1113
4    1113
1    1105
3    1097
Name: count, dtype: int64

<h1>Q25. Which month has the most serious accidents?</h1>

In [92]:
accident[accident['Accident_Severity'] == 'Serious'] ['Accident Date'].dt.month.value_counts()

Accident Date
9     7987
10    7939
6     7916
7     7863
5     7721
11    7568
8     7532
4     7336
3     7118
1     6681
12    6310
2     6246
Name: count, dtype: int64